syntax = "proto3";

package com.webank.eggroll.core.meta;

import "google/protobuf/duration.proto";

service RendezvousStore {
  rpc Set (SetRequest) returns (SetResponse) {}
  rpc Get (GetRequest) returns (GetResponse) {}
  rpc Add (AddRequest) returns (AddResponse) {}
  rpc CompareAndSet (CompareAndSetRequest) returns (CompareAndSetResponse) {}
  rpc Wait (WaitRequest) returns (WaitResponse) {}
  rpc NumKeys (NumKeysRequest) returns (NumKeysResponse) {}
  rpc DeleteKey (DeleteKeyRequest) returns (DeleteKeyResponse) {}
}


message SetRequest {
  string prefix = 1;
  bytes key = 2;
  bytes value = 3;
}

message SetResponse {}

message GetRequest {
  string prefix = 1;
  bytes key = 2;
  google.protobuf.Duration timeout = 3;
}

message GetResponse {
  bytes value = 1;
}

message AddRequest {
  string prefix = 1;
  bytes key = 2;
  int64 amount = 3;
}

message AddResponse {
  int64 amount = 1;
}

message CompareAndSetRequest {
  string prefix = 1;
  string key = 2;
  string expect = 3;
  string update = 4;
}

message CompareAndSetResponse {}

message WaitRequest {
  string prefix = 1;
  string key = 2;
  google.protobuf.Duration timeout = 3;
}

message WaitResponse {}

message NumKeysRequest {
  string prefix = 1;
}
message NumKeysResponse {
  int64 numKeys = 1;
}

message DeleteKeyRequest {
  string prefix = 1;
  string key = 2;
}

message DeleteKeyResponse {
  bool success = 1;
}